### PR TITLE
test/integration: clean up hanging 'threads'

### PR DESCRIPTION
**Pull request by alxndrsn**
Resolves need for mocha's `--exit` flag, which the docs say:

> enables a set of potential problems; it’s indicative of tests (or fixtures,
> harnesses, code under test, etc.) which don’t clean up after themselves
> properly. Ultimately, "dirty" tests can (but not always) lead to _false
> positive_ or _false negative_ results.

Flag originally introduced in:

* 472383e14ca6a7994cfd7ae9f563ebefbf74a0c1
* ca08b535f52a5d712d5c5f5544dcbda9002ec550

Problems resolved:

1. shut down db pool cleanly
2. avoid `yauzl` streams in cases where they hang

### CODE REVIEW

**Comment by alxndrsn** [comment]
I think this indicates another subtle bug in streams, probably in one of:

* the cleanup code in `test/util/zip.js`, or
* an unfulfilled interaction that `supertest` is expecting when working with streams, or
* a bug in `supertest`'s pipe/stream implementation

There's a very similar-looking issue filed at https://github.com/ladjs/supertest/issues/487.

**Comment by alxndrsn** [comment]
Confirmed these changes are still effective against current master:

# `db.end()`

## `master`, with no changes:

This demonstrates the need for an explicit call to `db.end()`.

```
$ time BCRYPT=insecure npx mocha --recursive test/integration --exit


  task: purge deleted resources (forms and submissions)
    forms
      force flag
        ✔ should not purge recently deleted forms by default
        ✔ should purge recently deleted form if forced
        ✔ should return message for multiple forms purged


  3 passing (3s)


real	0m4.343s
user	0m2.678s
sys	0m0.346s
```

vs.

```
$ time BCRYPT=insecure npx mocha --recursive test/integration


  task: purge deleted resources (forms and submissions)
    forms
      force flag
        ✔ should not purge recently deleted forms by default
        ✔ should purge recently deleted form if forced
        ✔ should return message for multiple forms purged


  3 passing (3s)


real	0m9.608s
user	0m2.858s
sys	0m0.393s
```

## `master` with `db.end()` added

```
$ time BCRYPT=insecure npx mocha --recursive test/integration --exit


  task: purge deleted resources (forms and submissions)
    forms
      force flag
        ✔ should not purge recently deleted forms by default
        ✔ should purge recently deleted form if forced
        ✔ should return message for multiple forms purged


  3 passing (3s)


real	0m4.495s
user	0m2.625s
sys	0m0.383s
```

vs

```
$ time BCRYPT=insecure npx mocha --recursive test/integration


  task: purge deleted resources (forms and submissions)
    forms
      force flag
        ✔ should not purge recently deleted forms by default
        ✔ should purge recently deleted form if forced
        ✔ should return message for multiple forms purged


  3 passing (3s)


real	0m4.528s
user	0m2.608s
sys	0m0.370s
```

# zip stream changes

## `master`, with `db.end()` included; no changes to zip stream code

```
$ time BCRYPT=insecure npx mocha --recursive test/integration --exit


  api: /forms/:id/submissions
    .csv.zip GET
::ffff:127.0.0.1 - - [08/Oct/2024:06:42:04 +0000] "POST /v1/sessions HTTP/1.1" 200 228
::ffff:127.0.0.1 - - [08/Oct/2024:06:42:04 +0000] "GET /v1/projects/1/forms/simple/submissions.csv.zip HTTP/1.1" 200 -
      ✔ should return the csv header even if there is no data (230ms)


  1 passing (3s)


real	0m4.806s
user	0m3.004s
sys	0m0.440s

```

vs.

```
$ time BCRYPT=insecure npx mocha --recursive test/integration


  api: /forms/:id/submissions
    .csv.zip GET
::ffff:127.0.0.1 - - [08/Oct/2024:06:42:10 +0000] "POST /v1/sessions HTTP/1.1" 200 228
::ffff:127.0.0.1 - - [08/Oct/2024:06:42:10 +0000] "GET /v1/projects/1/forms/simple/submissions.csv.zip HTTP/1.1" 200 -
      ✔ should return the csv header even if there is no data (223ms)

[SO FAR HANGING FOR 5 MINUTES]
```

## `master`, with `db.end()` and zip stream changes included

```
$ time BCRYPT=insecure npx mocha --recursive test/integration --exit


  api: /forms/:id/submissions
    .csv.zip GET
::ffff:127.0.0.1 - - [08/Oct/2024:06:49:37 +0000] "POST /v1/sessions HTTP/1.1" 200 228
::ffff:127.0.0.1 - - [08/Oct/2024:06:49:37 +0000] "GET /v1/projects/1/forms/simple/submissions.csv.zip HTTP/1.1" 200 -
      ✔ should return the csv header even if there is no data (205ms)


  1 passing (3s)


real	0m4.613s
user	0m2.847s
sys	0m0.417s
```

vs.

```
$ time BCRYPT=insecure npx mocha --recursive test/integration


  api: /forms/:id/submissions
    .csv.zip GET
::ffff:127.0.0.1 - - [08/Oct/2024:06:49:53 +0000] "POST /v1/sessions HTTP/1.1" 200 228
::ffff:127.0.0.1 - - [08/Oct/2024:06:49:53 +0000] "GET /v1/projects/1/forms/simple/submissions.csv.zip HTTP/1.1" 200 -
      ✔ should return the csv header even if there is no data (213ms)


  1 passing (3s)


real	0m4.542s
user	0m2.880s
sys	0m0.374s
```

File: Makefile
Comment (matthew-white) [inline Makefile:1]: I see other examples of `--exit` in this file, e.g., `test-fast` and `test-ci`.

File: test/util/zip.js [hunk 1]
```diff
@@ -12,6 +12,44 @@ const streamTest = require('streamtest').v2;
...
 //      …
 // }
+const processZipFile = (zipfile, callback) => {
+  const result = { filenames: [] };
+  const entries = [];
+  let completed = 0;
+
+  zipfile.on('entry', (entry) => entries.push(entry));
+  zipfile.on('end', (err) => {
+    if (err) return callback(err);
```
Comment (matthew-white) [inline test/util/zip.js:22]: More a note-to-self, previously `zipfile.on('end')` didn't check for an error passed to it.

File: test/util/zip.js [hunk 2]
```diff
@@ -23,50 +61,42 @@ const zipStreamToFiles = (zipStream, callback) => {
...
+// * the cleanup code in test/util/zip.js, or
+// * an unfulfilled interaction that supertest is expecting when working with streams, or
+// * a bug in supertest's pipe/stream implementation
+//
+// Possibly: https://github.com/ladjs/supertest/issues/487
+const binaryParser = (res, callback) => {
+  res.setEncoding('binary');
+  let data = '';
+  res.on('data', (chunk) => { data += chunk; });
+  res.on('end', () => {
```
Comment (matthew-white) [inline test/util/zip.js:83]: Is an error possible here / in this function? I don't see any error handling, but maybe errors aren't possible here.

File: test/util/zip.js [hunk 3]
```diff
@@ -23,50 +61,42 @@ const zipStreamToFiles = (zipStream, callback) => {
...
+  zipHttpResponse.buffer().parse(binaryParser).end((err, res) => {
+    if (err) return reject(err);
 
-module.exports = { zipStreamToFiles, pZipStreamToFiles };
+    // eslint-disable-next-line no-shadow
+    yauzl.fromBuffer(res.body, (err, zipfile) => {
+      if (err) return reject(err);
+
+      // eslint-disable-next-line no-shadow
+      processZipFile(zipfile, (err, result) => { if (err) reject(err); else resolve(result); });
```
Comment (matthew-white) [inline test/util/zip.js:97]: Would you mind making the `if; else;` two lines or preferably four? I'm not ready yet for both on one line. 😅

**Review by matthew-white — APPROVED** [review]

File: test/util/zip.js (hunk 1 above)
Comment (alxndrsn) [reply test/util/zip.js:22]: Maybe that was the entire problem :thinking: I'll check

File: test/util/zip.js (hunk 2 above)
Comment (alxndrsn) [reply test/util/zip.js:83]: I can't find clear docs on this, but I think this is an example of a `superagent` custom parser, mentioned at https://github.com/ladjs/superagent/blob/master/docs/index.md#parsing-response-bodies

The stock parsers are at https://github.com/ladjs/superagent/tree/2fd631ae225f4335fef37e9888925db0ef42d497/src/node/parsers, and don't seem to take an arg for the `end` event.

File: test/util/zip.js (hunk 1 above)
Comment (alxndrsn) [reply test/util/zip.js:22]: That wasn't the problem.  Maybe that `err` argument can be removed from the new code.

Comment (matthew-white) [reply test/util/zip.js]: Either way, probably doesn't hurt to keep it.

**Pull request merged**